lib/Gnuplot/Builder.pm

```diff
@@ -294,6 +294,23 @@ Arrange dataset options in valid order.
     $dataset = gfile('data.txt', with => "linespoints", using => "1:3"); ## NG!
     $dataset = gfile('data.txt', using => "1:3", with => "linespoints"); ## OK
 
+=item *
+
+Encode wide characters appropriately.
```

Comment:
What about this alternate wording?
```
=item *

Explicitly setting a character encoding.

When sending data to the external Gnuplot process,
L<Gnuplot::Builder::Process> uses whatever Perl thinks is the default
character encoding. In specific cases, it may be necessary to set a
specific encoding explicitly. See also L<perlunitut>.
```

Reply:
I think this kind of explanation should go to the document of `Gnuplot::Builder::Process/$ENCODING`. I'd like to make the document of `Gnuplot::Builder` as concise as possible.

Plus, I'm not sure about the following sentence.

> L\<Gnuplot::Builder::Process\> uses whatever Perl thinks is the default character encoding

In my understanding, by default (i.e. if `$ENCODING == undef`) `Gnuplot::Builder::Process` does NOTHING about encoding scripts. It does not use "the default character encoding" or whatever.

Reply:
> I think this kind of explanation should go to the document of `Gnuplot::Builder::Process/$ENCODING`. I'd like to make the document of `Gnuplot::Builder` as concise as possible.

Where would you move it?

> Plus, I'm not sure about the following sentence.
> 
> > L<Gnuplot::Builder::Process> uses whatever Perl thinks is the default character encoding
> 
> In my understanding, by default (i.e. if `$ENCODING == undef`) `Gnuplot::Builder::Process` does NOTHING about encoding scripts. It does not use "the default character encoding" or whatever.

Perl uses an internal string representation. Whenever data from the outside world is stored in a Perl string, or when data from a Perl string is to be sent to the outside world, decoding and encoding happen. See [perlunitut](https://perldoc.perl.org/perlunitut) for more details.

Hence, while `Gnuplot::Builder::Process` does not perform or trigger any encoding itself, Perl and its IO layers will still perform encoding and decoding. 

For information about what "Perl thinks is the default character encoding", see [this bit of the PerlIO documentation](https://perldoc.perl.org/PerlIO#Defaults-and-how-to-override-them).

All of the above is true as of Perl 5.8.0, btw.

Reply:
> Where would you move it?

Here -> https://metacpan.org/pod/Gnuplot::Builder::Process#$ENCODING

> For information about what "Perl thinks is the default character encoding", see this bit of the PerlIO documentation.

Thanks for the pointer. I've never used PerlIO features seriously, so I was surprised that there were many ways to configure it.

> Hence, while Gnuplot::Builder::Process does not perform or trigger any encoding itself, Perl and its IO layers will still perform encoding and decoding.

In theory, it's true. However, in practice, I don't think it's possible to configure character encoding IO layers on the channel between `Gnuplot::Builder` and a gnuplot process. Here's why.

- The [PerlIO document](https://perldoc.perl.org/PerlIO#Defaults-and-how-to-override-them) says the default IO layers don't do any character encoding/decoding jobs.
- You can configure [PERLIO environment variable](https://perldoc.perl.org/perlrun#PERLIO) to modify the default IO layers. However, it cannot accept character encoding/decoding layers.
- You can use [open pragma](https://perldoc.perl.org/open) to configure IO layers for handles made by `open()` function. However, the pragma's effect is limited in the lexical scope. You cannot set IO layers for a gnuplot process unless you directly modify the code of `Gnuplot::Builder::Process`.
- You can use [PERL_UNICODE environment variable](https://perldoc.perl.org/perlrun#PERL_UNICODE) or [-C command-line option](https://perldoc.perl.org/perlrun#-C-%5Bnumber/list%5D) to configure IO layers. However, their effect is limited in the main program scope. They have no effect on code run in modules.

So, I concluded that there is no way for a user of `Gnuplot::Builder` to configure character encoding/decoding IO layers on the channel between `Gnuplot::Builder` and a gnuplot process. If you think there is a way, just give me a sample code.

So, while the sentence around "the default character encoding" is true in theory, I think it is too complicated and confusing.

Reply:
> > Where would you move it?
> 
> Here -> https://metacpan.org/pod/Gnuplot::Builder::Process#$ENCODING

I see. Yes, that would be a good place.

> [...]
> > Hence, while Gnuplot::Builder::Process does not perform or trigger any encoding itself, Perl and its IO layers will still perform encoding and decoding.
> 
> In theory, it's true. However, in practice, I don't think it's possible to configure character encoding IO layers on the channel between `Gnuplot::Builder` and a gnuplot process. Here's why.
> [...]
> So, I concluded that there is no way for a user of `Gnuplot::Builder` to configure character encoding/decoding IO layers on the channel between `Gnuplot::Builder` and a gnuplot process. If you think there is a way, just give me a sample code.

I get the impression that you think I were claiming `$ENCODING` was not needed, and other ways of getting the same effect would exist?

If that is the case, then please accept my apologies for not being clear enough. `$ENCODING` is needed to do exactly what it does. No doubt abut that.

Hence, from my point of view we are only discussing how much to say about it in the documentation.

> So, while the sentence around "the default character encoding" is true in theory, I think it is too complicated and confusing.

As is explained in [the docs I linked to earlier](https://perldoc.perl.org/PerlIO#Defaults-and-how-to-override-them), there is always some default encoding in place, which is determined when Perl is being built. Hence, even if you don't specify anything, Perl will decode and encode text upon input and output using the "natural" format of the platform for which Perl was built.

To cut a long story short: I think it would be fine to leave the comment, with which this discussion started, as is.

Instead, I would propose to apply a small modification to the documentation for `$ENCODING` in `Process.pm`, which now reads:

> $ENCODING
> 
> If set, [Gnuplot::Builder](https://metacpan.org/pod/Gnuplot::Builder) encodes the script string in the specified encoding just before streaming into the gnuplot process. You can specify any encoding names recognizable by [Encode](https://metacpan.org/pod/Encode) module.
> 
> By default it's undef, meaning it doesn't encode the script.
> 
> You can also set this variable by the environment variable PERL_GNUPLOT_BUILDER_PROCESS_ENCODING.

Now my proposal would be to change
> By default it's undef, meaning it doesn't encode the script.

to

> By default it's undef.

Looking forward to your toughts.

Reply:
> As is explained in the docs I linked to earlier, there is always some default encoding in place,

> Hence, even if you don't specify anything, Perl will decode and encode text upon input and output using the "natural" format of the platform for which Perl was built.

I'm still not sure if the above two sentences are correct.

I read [the document you linked](https://perldoc.perl.org/PerlIO#Defaults-and-how-to-override-them), but couldn't find anything that indicates such a default encoding layer. The document said that the default layers can include `:unix`, `:crlf`, `:stdio`, `:perlio` (depending on the platform), but I think none of them does encoding/decoding jobs.

Exactly which part of the document do you think indicates such a default encoding layer?

Of course, it can depend on how the perl runtime is built. So, I did some experiments.

I saved the following perl script (encoded in UTF-8) as `exp_perlio.pl`

```perl
use strict;
use warnings;
use utf8;

my $s = "にほんご\n";  ## or whatever string that includes non-ASCII characters

print $s;
```

and ran it. (The `env -i` clears all environment variables)

```
$ env -i perl exp_perlio.pl 
Wide character in print at exp_perlio.pl line 7.
にほんご
```

The `Wide character...` warning indicates that the script directly writes an "internal string representation" to STDOUT without encoding. That means there is no encoding/decoding layer working at STDOUT.

I can set a UTF-8 encoding/decoding layer at STDOUT by a command-line switch. 

```
$ env -i perl -CO exp_perlio.pl 
にほんご
```

Now that the layer enabled by `-CO` encodes the internal string representation into UTF-8, the warning has gone.

The output of `env -i perl -V` is here: [perlV.txt](https://github.com/debug-ito/Gnuplot-Builder/files/13210921/perlV.txt)


So, could you run the same experiment, and put your result on `env -i perl -V`? That way, I can see how different perl runtimes have different settings on PerlIO layers.

Reply:
As for your suggestion,

> By default it's undef.

I don't think it's good. It doesn't explain the meaning of setting it to `undef`. Probably I should add some extra explanation to it, rather than removing something.

Reply:
> [...]
> I read [the document you linked](https://perldoc.perl.org/PerlIO#Defaults-and-how-to-override-them), but couldn't find anything that indicates such a default encoding layer. The document said that the default layers can include `:unix`, `:crlf`, `:stdio`, `:perlio` (depending on the platform), but I think none of them does encoding/decoding jobs.
> 
> Exactly which part of the document do you think indicates such a default encoding layer?

At the very end of "Defaults and how to override them" it summarises the default behaviour:

**--- BEGIN QUOTE ---**
The following table summarizes the default layers on UNIX-like and DOS-like platforms and depending on the setting of $ENV{PERLIO}:
```
PERLIO     UNIX-like                   DOS-like
------     ---------                   --------
unset / "" :unix:perlio / :stdio [1]   :unix:crlf
:stdio     :stdio                      :stdio
:perlio    :unix:perlio                :unix:perlio

# [1] ":stdio" if Configure found out how to do "fast stdio" (depends
# on the stdio implementation) and in Perl 5.8, else ":unix:perlio"
```
**--- END QUOTE ---**

Here, the `unset` line is what applies when `$ENCODING` is `undef`.

> Of course, it can depend on how the perl runtime is built. So, I did some experiments.
> [...]
> ```
> $ env -i perl exp_perlio.pl 
> Wide character in print at exp_perlio.pl line 7.
> にほんご
> ```
> 
> The `Wide character...` warning indicates that the script directly writes an "internal string representation" to STDOUT without encoding. That means there is no encoding/decoding layer working at STDOUT.
> [...]

No, that is not strictly true. The warning indicates that Perl detected something in the string, that may or may not produce the desired result with the default input/output processing as per the table I quoted above. It does **not** indicate that no encoding is happening. It simply says that the default encoding may not be what you wanted.

In your case, luckily, the default encoding produced the expected result as both, the editor used for your script and the (pseudo) tty used to run your experiment happened to use the same encoding (which is no surprise as they were built for one and the same platform).

To shed more light on Perl's behaviour, I further analysed the output data produced by your examples:
```console
foo@bar:~$ env -i perl exp_perlio.pl | od -x
Wide character in print at exp_perlio.pl line 7.
0000000      81e3    e3ab    bb81    82e3    e393    9481    000a
0000015
```

Now with the "correct" output encoding:
```console
foo@bar:~$ env -i perl -CO exp_perlio.pl | od -x
0000000      81e3    e3ab    bb81    82e3    e393    9481    000a
0000015
```

The actual sequence of bytes produced is **exactly** the same in both cases, and on my platform apparently UTF-8. The only visible thing that the `-CO` switch changed was to suppress the warning.

Thus the warning is **not** to say that no encoding were happening, but it is rather to inform the user that Perl did not have specific instructions on what to do in this case, and that therefore the results may not be what the user expected.

In case you feel like more research, additional background can be found in the W3C articles [here](https://www.w3.org/International/questions/qa-what-is-encoding) and [here](https://www.w3.org/International/articles/definitions-characters/), and in the [Wikipedia entry on character encoding](https://en.wikipedia.org/wiki/Character_encoding). In case you prefer a denser, more technically minded presentation, you could look to the [Unicode® Technical Report #17 "Unicode Character Encoding Model"](https://www.unicode.org/reports/tr17/).

Reply:
Thanks for explanation.

Would you run the same experiment and share the result with me?